### ISSUE TITLE
Floating control text becomes smaller and smaller.

### ISSUE THREAD

**Issue by pyscripter** [SilverpointDev/sptbxlib#68]
Steps to reproduce:

- Run Overview in a high DPI monitor.
- Drag the toolbar containg "Address" out of the Dock to make it floating without releasing the mouse.
- With the mouse pressed put the toolback in the original Dock and then out again.
-  Repeat the previous step without releasing the mouse.

Result:  The control  text becomes smaller and smaller.

If you release the mouse after docking back the toolbar the correct text size is shown.

The issue is reproducible with TB2K without SpTBX.

**Comment by pyscripter** [comment]
I have updated the TB2K patch to fix this error.  Please see the attached.

[tb2kdiff.txt](https://github.com/SilverpointDev/sptbxlib/files/5759650/tb2kdiff.txt)

**Comment by SilverpointDev** [comment]
Thank you for the time you spent testing the lib.
After applying the changes the combo shrinks one time when you float the toolbar.

Question, on TTBCustomDockableWindow.SetParent you call ScaleForPPI:
      // Normally, setting the Parent scales the Form but here it doesn't because
      // CustomDocakbleWindow has a FreeNotification (see TControl.SetParent)
If you comment ScaleForPPI calls the DockableWindow is still scaled.

If I comment ScaleForPPI calls and comment the change you made on TBCustomDockableWindow.SetFloating then it works as expected:
//            Font.Height := MulDiv(Font.Height, 96, Screen.PixelsPerInch);

**Comment by pyscripter** [comment]
> If I comment ScaleForPPI calls and comment the change you made on TBCustomDockableWindow.SetFloating then it works as expected:

I have a primary monitor with DPI = 144 and a secondary monitor with PPI 192.  If I do the above and move the form on the secondary monitor the font is to high.

> After applying the changes the combo shrinks one time when you float the toolbar.
The shrinkage is probably due to the different font used in the floating window.  I have kept SetParent as in the patch and updated the SetFloating to:

```pascal
procedure TTBCustomDockableWindow.SetFloating(Value: Boolean);
var
  ParentFrm: TTBCustomForm;
  NewFloatParent: TTBFloatingWindowParent;
begin
  if FFloating <> Value then begin
    if Value and not(csDesigning in ComponentState) then begin
      ParentFrm := TBValidToolWindowParentForm(Self);
      if FFloatParent = nil then begin
        NewFloatParent := GetFloatingWindowParentClass.CreateNew(nil);
        try
          with NewFloatParent do begin
            // The form font is initialized with the font size of the main monitor
            // If the main monitor has PPI > 96 and we then scale
            // child controls with ParentFont property would have wrong font size
            // after we scale the form.  Hence the following:
            ParentFont := False;
            Font.Assign(Screen.MenuFont); // same as ToolbarFont
            Font.Height := MulDiv(Font.Height, 96, Screen.PixelsPerInch);
            FDockableWindow := Self;
            BorderStyle := bsToolWindow;
            ShowHint := True;
            Visible := True;

            { Note: The above line doesn't actually make it visible at this
              point since FShouldShow is still False. }
          end;
        except
          NewFloatParent.Free;
          raise;
        end;
        FFloatParent := NewFloatParent;
      end;
      ParentFrm.FreeNotification(FFloatParent);
      FFloatParent.FParentForm := ParentFrm;
      FFloatParent.Caption := Caption;
      Parent := FFloatParent;
      SetBounds(0, 0, Width, Height);
    end
    else
      Parent := TBValidToolWindowParentForm(Self);
  end;
end;
```

Here floating works well under all scenarios.  Even when you move the toolbar  to another monitor, double-click to dock it and double-click to float it.

**Comment by pyscripter** [comment]
By the way you can delete the following in SpTBXExtEditors.pas

` TSpTBXPreviewPanel = class(TPanel);`

**Comment by pyscripter** [comment]
Let me provided some explanations of the patched code.

** Floating form scaling **
Forms that are created manually (not loaded from resources) are not scaled automatically (FCurrentPPI = 96).   
If/when they are shown in a secondary monitor they get scaled via TCustomForm.WMDpiChanged.  However if the primary monitor DPI <> 96. the font size will be wrong (bigger).  This is because the font is initialized with the primary monitor font size, but then scaled from 96 to the secondary monitor DPI.

**  Parented control  (toolbar) scaling **
When you set as a parent of the toolbar the floating window no scaling takes place because of the notification.

The code in TBDock.pas I have added makes sure:
- Font size is set at the proper 96 dpi font height and not the primary monitor font height
- The floating window is scaled to the destination monitor dpi .  It may be scaled twice if shown in the secondary monitor but this does no harm (The second call will do nothing).
- The parented control (toolbar) is scaled at the destination monitor dpi before the new parent is set.

So with the above both the floating window and what it contains will be scaled at the right dpi.  If you do none of the above, floating toolbars will work when you float the toolbar on the same monitor (almost by accident) but it will not be properly drawn in other scenarios.

**Comment by SilverpointDev** [comment]
Adding all your changes I get a resizing of the floating toolbar controls.

On a 96 dpi monitor
1) Drop a dock, a toolbar and a TSpTBXCombobox inside the toolbar.
2) Run, the combo is 21 of height.
3) Double click the toolbar to make it float, the combo is 23 of height and the toolbar borders are painted incorrectly due to an incorrect item calc sizing.

On TTBCustomDockableWindow.SetFloating:
```pascal
            ParentFont := False;
            Font.Assign(Screen.MenuFont); // same as ToolbarFont
            Font.Height := MulDiv(Font.Height, 96, Screen.PixelsPerInch);
```

I get different fonts after this, even when the DPI is not changed.
Before 
Name: Tahoma, Size: 8, Height: -11

After:
Name: Segoe UI, Size: 9, Height: -12

Changing Screen.MenuFont to Self.Font seems to solve the problem:
  Font.Assign(Screen.MenuFont); // same as ToolbarFont
To
  Font.Assign(Self.Font);

**Comment by pyscripter** [comment]
> Font.Assign(Self.Font);

I have used:

```pascal
            Font.Assign(ParentFrm.Font);
            Font.Height := MulDiv(Font.Height, 96, ParentFrm.CurrentPPI);
```
So that the font is the same as in the original font.

I have simplified the scaling.   Please see attached new patch.  Please let me know whether it works.
[tb2kdiff.txt](https://github.com/SilverpointDev/sptbxlib/files/5792636/tb2kdiff.txt)

**Comment by pyscripter** [comment]
@SilverpointDev I wonder whether you had a chance to test the latest patch above...

**Comment by SilverpointDev** [comment]
Seems to work fine now

**Comment by pyscripter** [comment]
Great!  In that case could you please update the patch in the repo based on the above and including your other fix from #71.

You can then close this issue.

**Comment by pyscripter** [comment]
I see that you have done that.  So I am closing the issue.
